test-data/unit/fixtures/typing-typeddict.pyi

```diff
@@ -71,3 +71,18 @@ class _TypedDict(Mapping[str, object]):
     def pop(self, k: NoReturn, default: T = ...) -> object: ...
     def update(self: T, __m: T) -> None: ...
     def __delitem__(self, k: NoReturn) -> None: ...
+    # It is a bit of a lie:
+    # 1. it is only supported since 3.9 in runtime
+    # 2. `__or__` uses `dict[str, Any]`, not `Mapping[str, object]`
+    @overload
+    def __or__(self, __value: Self) -> Self: ...
+    @overload
+    def __or__(self, __value: Mapping[str, object]) -> Mapping[str, object]: ...
+    # TODO: re-enable after `__ror__` definition is fixed
+    # https://github.com/python/typeshed/issues/10678
+    # @overload
+    # def __ror__(self, __value: Self) -> Self: ...
+    # @overload
+    # def __ror__(self, __value: dict[str, Any]) -> dict[str, object]: ...
+    # supposedly incompatible definitions of __or__ and __ior__
```

Comment:
I don't understand this comment: see what I wrote in https://github.com/python/typeshed/issues/10678#issuecomment-1759395609

Reply:
Here's my understanding of this problem:
1. `dict.__or__` does not allow `TypedDict`, because it only allows `dict` (and `TypedDict` does not match it):

```
ex.py:7: error: No overload variant of "__or__" of "dict" matches argument type "Foo"  [operator]
ex.py:7: note: Possible overload variants:
ex.py:7: note:     def __or__(self, dict[str, int], /) -> dict[str, int]
ex.py:7: note:     def [_T1, _T2] __or__(self, dict[_T1, _T2], /) -> dict[str | _T1, int | _T2]
ex.py:7: note: Revealed type is "Any"
ex.py:8: error: No overload variant of "__or__" of "dict" matches argument type "Foo"  [operator]
ex.py:8: note: Possible overload variants:
ex.py:8: note:     def __or__(self, dict[str, str], /) -> dict[str, str]
ex.py:8: note:     def [_T1, _T2] __or__(self, dict[_T1, _T2], /) -> dict[str | _T1, str | _T2]
ex.py:8: note: Revealed type is "Any"
```

2. mypy does the reverse checking for every `__or__` operator, so even for `foo | {'key': 1}` it raises (without this PR):

```
ex.py:7: error: No overload variant of "__ror__" of "dict" matches argument type "Foo"  [operator]
ex.py:7: note: Possible overload variants:
ex.py:7: note:     def __ror__(self, dict[str, int], /) -> dict[str, int]
ex.py:7: note:     def [_T1, _T2] __ror__(self, dict[_T1, _T2], /) -> dict[str | _T1, int | _T2]
ex.py:7: note: Revealed type is "Any"
```

So, my problem definition is: `dict.__or__` is not compatible with `TypedDict` and `dict.__ror__` is also not compatible with `TypedDict`

Reply:
Here is the snippet the OP was complaining about in #16244:

```py
from typing import TypedDict

class Foo(TypedDict):
    key: int
    
foo: Foo = {"key": 1}
foo |= {"key": 3}
```

mypy reports:

```
error: Incompatible types in assignment (expression has type "dict[str, object]", variable has type "Foo")  [assignment]
```

https://mypy-play.net/?mypy=latest&python=3.11&gist=9927e1732f6fce4595d29a19493ae603

Note: mypy does not report that `|=` cannot be used between `foo` and `{"key: 3}`. It understands that the `|=` operator _can_ be used between the two objects. However, it is inferring that the type of the `foo` variable after the `|`-ing is `dict[str, object]` rather than `Foo`, which would be an incompatible reassignment of the `foo` variable, since the `foo` variable has been declared to always have type `Foo`.

That indicates that mypy is using the second overload of `_TypedDict.__or__` here to infer the type of `foo` after the `|=` operation, rather than `_TypedDict.__ior__` or the first overload of `_TypedDict.__or__`. If it used the `_TypedDict.__ior__` annotations, or the first overload of `_TypedDict.__or__`, it would be able to correctly infer that the type of `foo` after the `|=` operation is still `Foo`, which is a reassignment that's compatible with the declared type of the `foo` variable:

https://github.com/python/mypy/blob/2e52e98fd2873775a58616c097e93c96f58fc991/mypy/typeshed/stdlib/typing.pyi#L910-L919

I think the issue is that in this snippet, mypy is not able currently to infer using type context (bidirectional type inference) that the object on the right-hand-side of the `|=` operator can be considered to be an instance of `Foo`. If it did so, then it would select `_TypedDict.__ior__` or the first overload of `_TypedDict.__or__` to infer the type of `foo` after the `|=` operation, and would correctly understand that the type of the `foo` variable does not change as a result of the operation.